pandas/indexes/base.py

```diff
@@ -966,44 +968,41 @@ def _convert_scalar_indexer(self, key, kind=None):
         Parameters
         ----------
         key : label of the slice bound
-        kind : optional, type of the indexing operation (loc/ix/iloc/None)
-
-        right now we are converting
+        kind : {'ix', 'loc', 'getitem', 'iloc'} or None
         """
 
+        assert kind in ['ix', 'loc', 'getitem', 'iloc', None]
+
         if kind == 'iloc':
-            if is_integer(key):
-                return key
-            return self._invalid_indexer('positional', key)
-        else:
+            return self._validate_indexer('positional', key, kind)
 
-            if len(self):
-
-                # we can safely disallow
-                # if we are not a MultiIndex
-                # or a Float64Index
-                # or have mixed inferred type (IOW we have the possiblity
-                # of a float in with say strings)
-                if is_float(key):
-                    if not (isinstance(self, ABCMultiIndex,) or
-                            self.is_floating() or self.is_mixed()):
-                        return self._invalid_indexer('label', key)
-
-                # we can disallow integers with loc
-                # if could not contain and integer
-                elif is_integer(key) and kind == 'loc':
-                    if not (isinstance(self, ABCMultiIndex,) or
-                            self.holds_integer() or self.is_mixed()):
-                        return self._invalid_indexer('label', key)
+        if len(self) and not isinstance(self, ABCMultiIndex,):
 
-        return key
+            # we can raise here if we are definitive that this
+            # is positional indexing (eg. .ix on with a float)
+            # or label indexing if we are using a type able
+            # to be represented in the index
 
-    def _convert_slice_indexer_getitem(self, key, is_index_slice=False):
-        """ called from the getitem slicers, determine how to treat the key
-            whether positional or not """
-        if self.is_integer() or is_index_slice:
-            return key
-        return self._convert_slice_indexer(key)
+            if kind in ['getitem', 'ix'] and is_float(key):
```

Comment:
This is where it aises a ``TypeError`` here if its label based but not the right type.